File: tks/RepositoriesAdapters/src/test/kotlin/repository/repositories/UserRepositoryTest.java
package repository.repositories;

import repository.data.UserEntity;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

public class UserRepositoryTest {

    RepositoryBase<UserEntity> repository;
    List<UserEntity> store;
    UserEntity sampleUser1;
    UserEntity sampleUser2;

    @BeforeEach
    public void init(){
        sampleUser1 = new UserEntity("mszewc@edu.pl", "ADMIN", "####", true);
        sampleUser2 = new UserEntity("mzab@edu.pl", "CLIENT", "####", true);
        store = new ArrayList<>();
        repository = new RepositoryBase<>(store);
    }

    @Test
    public void AddNewUserWithSetId_ShouldFail(){
        sampleUser1.setGuid(UUID.randomUUID());
        assertThrows(Exception.class, () -> repository.add(sampleUser1));
        assertEquals(0, store.size());
    }

    @Test
    public void AddDuplicatedUser_ShouldFail(){
        repository.add(sampleUser1);
        sampleUser2.setGuid(sampleUser1.getGuid());
        assertThrows(Exception.class, () -> repository.add(sampleUser2));
        assertEquals(1, store.size());
        assertNotSame(sampleUser2, store.get(0));
    }

    @Test
    public void UpdateNonExistingEntity_ShouldFail(){
        assertThrows(Exception.class, () -> repository.update(sampleUser1));
        sampleUser1.setGuid(UUID.randomUUID());
        assertThrows(Exception.class, () -> repository.update(sampleUser1));
        repository.add(sampleUser2);
        assertThrows(Exception.class, () -> repository.update(sampleUser1));
    }

    @Test
    public void UpdateValidUser_ShouldSuccess(){
        repository.add(sampleUser1);
        assertSame(sampleUser1, store.get(0));
        sampleUser2.setGuid(sampleUser1.getGuid());
        repository.update(sampleUser2);
        assertSame(sampleUser2, store.get(0));
    }

    @Test
    public void RemoveNonExistingEntity_ShouldFail(){
        sampleUser1.setGuid(UUID.randomUUID());
        assertThrows(Exception.class, () -> repository.remove(sampleUser1));
    }

    @Test
    public void FindTests(){
        var res1 = repository.find(user -> user.getEmail().equals("mszewc@edu.pl"));
        assertNull(res1);
        repository.add(sampleUser1);
        var res2 = repository.find(user -> user.getEmail().equals("mszewc@edu.pl"));
        assertNotNull(res2);
        assertSame(sampleUser1, res2);
    }
}

File: tks/RepositoriesAdapters/src/test/kotlin/repository/repositories/UserRepositoryTest.kt
package repository.repositories

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import repository.data.UserEntity
import java.util.*

class UserRepositoryTest {
    private val store: MutableList<UserEntity> = ArrayList<UserEntity>()
    private val sampleUser1: UserEntity = UserEntity(null, "mszewc@edu.pl", "ADMIN", "####", true)
    private val sampleUser2: UserEntity = UserEntity(null, "mzab@edu.pl", "CLIENT", "####", true)
    private val repository: RepositoryBase<UserEntity> = RepositoryBase(store)

    @Test
    fun `Given entity with id, add should fail`() {
        sampleUser1.guid = UUID.randomUUID()
        Assertions.assertThrows(Exception::class.java) { repository.add(sampleUser1) }
        Assertions.assertEquals(0, store.size)
    }

    @Test
    fun `Given duplicated entity, add should fail`() {
        repository.add(sampleUser1)
        sampleUser2.guid = sampleUser1.guid
        Assertions.assertThrows(Exception::class.java) { repository.add(sampleUser2) }
        Assertions.assertEquals(1, store.size)
        Assertions.assertNotSame(sampleUser2, store[0])
    }

    @Test
    fun `Given nonexistent entity, update should fail`() {
        Assertions.assertThrows(Exception::class.java) { repository.update(sampleUser1) }
        sampleUser1.guid = UUID.randomUUID()
        Assertions.assertThrows(Exception::class.java) { repository.update(sampleUser1) }
        repository.add(sampleUser2)
        Assertions.assertThrows(Exception::class.java) { repository.update(sampleUser1) }
    }

    @Test
    fun `Given valid entity, update should success`() {
        repository.add(sampleUser1)
        Assertions.assertSame(sampleUser1, store[0])
        sampleUser2.guid = sampleUser1.guid
        Assertions.assertEquals(sampleUser1, sampleUser2)
        repository.update(sampleUser2)
        Assertions.assertSame(sampleUser2, store[0])
    }

    @Test
    fun `Given nonexistent entity, remove should fail`() {
        sampleUser1.guid = UUID.randomUUID()
        Assertions.assertThrows(Exception::class.java) { repository.remove(sampleUser1) }
    }

    @Test
    fun `Given valid predicate, fin should return entity`() {
        val res1 = repository.find { user: UserEntity? -> user!!.email == "mszewc@edu.pl" }
        Assertions.assertNull(res1)
        repository.add(sampleUser1)
        val res2 = repository.find { user: UserEntity? -> user!!.email == "mszewc@edu.pl" }
        Assertions.assertNotNull(res2)
        Assertions.assertSame(sampleUser1, res2)
    }
}